docs/internal/abi.md

```diff
@@ -0,0 +1,444 @@
+
+There are discussions on introducing non-power-of-two integers and other floating point sizes, but they won't be present for the first iteration of the compiler.
+
+### Pointers
+
+Pointers have different sizes depending on the target architecture:
+* 32-bit integers representing an address.
+* 64-bit integers representing an address.
+* 128-bit CHERI capabilities granting access to an address range.
+
+All Verona objects are at least 8-byte aligned and so the low three bits of a pointer will always be zero.
+When targeting a system that uses integer addresses, all pointers have the same representation.
+When targeting a CHERI system, the following additional rules apply:
+* Pointers to `cown`s are sealed with a type reserved for the runtime.
+  This ensures that `cown`s in immutable objects cannot be dereferenced by unsafe (sandboxed) code and provides some defense in depth against compiler bugs that would accidentally dereference a `cown` pointer.
+* Pointers to immutable objects are not represented by read-only capabilities.
+  Doing so would require expensive operations for reference-count manipulation.
+  When a pointer to an immutable object is passed via the foreign-code layer, it must have the load-storable-capabilities permission removed, such that no reachable capability will ever provide write permission.
+
+### Classes and interfaces
+
+Concrete classes and interfaces each have their own unique header, with a unique descriptor.
+Following a pointer and decoding the header gives you the type of the object and therefore there is no need to treat interfaces and classes differently.
+
+Classes and interfaces are stored similar to a _C structure_.
+The general layout is:
+* A header.
+* The list of fields (with `embed` values or pointers to objects).
+
+The header is a pair of values:
+* The region meta-data, a pointer-sized value containing information for the runtime library.
+* A pointer to the type descriptor (see below).
+
+Each object points to a descriptor that uniquely identifies its type.
+Pattern matching by type, when the type of the object is not statically known at compile time, involves the descriptor.
+Matching against a concrete type is a simple comparison: does the descriptor pointer in the object point to the descriptor for the concrete type?
+Matching a pointer to an unknown concrete type (i.e. whose static type is an interface type) against an interface type requires a more complex lookup.
+
+The remaining fields can, on internal representations, be packed or reordered for optimisation purposes.
+Machine-word types (ex. numeric) are always represented by value (their singleton representation).
+
+For example, a class:
+```ts
+class Other { ... }
+class Foo
+{
+    var a : U32 & imm;
+    embed var b : F64 & imm;
+    var other : Other & mut;
+    create() { ... }
+}
+```
+
+A naive layout could be:
+```ts
+ { { ptr, ptr }, i32, f64,  ptr   }
+      header,    U32, F64, Other*
```

Comment:
```suggestion
      header,      a,   b, other*
```

`embed` doesn't seem to be doing anything here, tho'.

Do you mean for that F64 to be unnaturally aligned on 64-bit machines, are you suppressing padding, should the example say "on a 32-bit machine", or am I missing something?

Reply:
Sorry, that was a bad example made worse. `embed` did something before my latest change but it was the wrong something. I'll try to come up with a better example.

I was not suggesting padding and used the term "naive" to show unaligned F64 on purpose. the paragraph below explains why.